Incrementally lex and parse data

wingo opened the pull request:
Instead of building an AST, then parsing that AST, incrementally parse
as we go.  This should prevent the creation of large intermediate data
structures when parsing 100-MB files that could push us over LuaJIT's 1G
limit.

Fixes #624; I can load the 1M-element table with this patch.  However it takes 4 minutes to compile the damn thing!!!!  Needs further optimization.

wingo committed:
Incrementally lex and parse data

Instead of building an AST, then parsing that AST, incrementally parse
as we go.  This should prevent the creation of large intermediate data
structures when parsing 100-MB files that could push us over LuaJIT's 1G
limit.

wingo committed:
Merge commit '4803bb039b91dccb9e6a732de6a0069d3a0ff907' into incremental-data-parse

wingo committed:
Micro-optimizations to data parser.

wingo committed:
Add a fast path for ctable insert.

wingo committed:
Use 40% occupancy as a maximum for data ctables.

wingo commented:
It seems that ctable insertion was one of the slow things -- this patchset changes the max occupancy rate for ctables to 0.4.  Come to think of it, I think that the migrated files are in hash order; does insertion in hash order have worse complexity, I wonder?

wingo commented:
We don't have the same problems as http://accidentallyquadratic.tumblr.com/post/153545455987/rust-hash-iteration-reinsertion because we don't modulo, but I wonder, probably we aren't hitting the fast paths as much as we should.

wingo commented:
Anyway, with this patchset, compiling the 1M-entry table works now, and it's down to 1m22s.  About three times as long as our custom parser used to do but oh well.

tsyesika reviewed (approved):
LGTM.

dpino commented on src/lib/ctable.lua:
```diff
@@ -219,9 +221,19 @@ function CTable:insert(hash, key, value, updates_allowed)
    local entries = self.entries
    local scale = self.scale
-   local start_index = hash_to_index(hash, self.scale)
+   -- local start_index = hash_to_index(hash, self.scale)
```
Nit: Remove commented line.

wingo replied:
I left it there so as to be greppable in case the function needs changing -- there are a few more instances of the function call in the file and I thought it wasn't so great to inline them all; dunno.

dpino commented:
I grasp the general picture but I don't fully understand how it works to be honest.:) I trust Jessica's judge. LGTM.

dpino commented:
Snabb-bot returned an error in lib.yang.util. It complains cannot find function `check`. It was removed in the patch.

```
$ sudo ./snabb snsh -t lib.yang.value
lib/yang/util.lua:33: variable 'check' is not declared
```

wingo committed:
Inline another call to check in tointeger

wingo committed:
Fix bug in ctable hash-to-index mapping

Before, we would sometimes map a hash beyond the size of a table.  It
didn't usually matter before as resizing always visited 2*size.  Now
that resize precisely visits only the active entries -- those within
[0,size+max_displacement) -- now we see this bug.  Caught by the cltable
test suite, yay!

dpino reviewed (commented):
LGTM
